modules/rpc/yarpc.go

```diff
@@ -36,66 +35,138 @@ import (
 	tch "go.uber.org/yarpc/transport/tchannel"
 )
 
-// YarpcModule is an implementation of a core module using YARPC
-type YarpcModule struct {
+// YARPCModule is an implementation of a core module using YARPC.
+// All the YARPC modules share the same dispatcher and middleware.
+type YARPCModule struct {
 	modules.ModuleBase
-	rpc                      *yarpc.Dispatcher
-	register                 registerServiceFunc
-	config                   yarpcConfig
-	log                      ulog.Log
-	stateMu                  sync.RWMutex
-	inboundMiddlewares       []middleware.UnaryInbound
-	onewayInboundMiddlewares []middleware.OnewayInbound
+	register  registerServiceFunc
+	config    yarpcConfig
+	log       ulog.Log
+	stateMu   sync.RWMutex
+	isRunning bool
 }
 
 var (
-	_dispatcherFn = defaultYarpcDispatcher
+	_dispatcherFn = defaultYARPCDispatcher
 
-	_ service.Module = &YarpcModule{}
+	_ service.Module = &YARPCModule{}
+
+	// Represents the global dispatcher used by all started modules.
+	_dispatcher   *yarpc.Dispatcher
+	_muDispatcher sync.RWMutex
+
+	_configsOnce sync.Once
+	_configs     configCollection
 )
 
-type registerServiceFunc func(module *YarpcModule)
+type registerServiceFunc func(module *YARPCModule)
 
 type yarpcConfig struct {
 	modules.ModuleConfig
-	Bind          string `yaml:"bind"`
-	AdvertiseName string `yaml:"advertiseName"`
+	Bind                    string `yaml:"bind"`
+	AdvertiseName           string `yaml:"advertiseName"`
+	inboundMiddleware       []middleware.UnaryInbound
+	onewayInboundMiddleware []middleware.OnewayInbound
+	inbounds                []transport.Inbound
+}
+
+// Stores a collection of all modules configs and provides
+// operations that are safe to call from multiple go routines.
+type configCollection struct {
+	sync.RWMutex
+	configs []*yarpcConfig
+}
+
+func (c *configCollection) addConfig(config *yarpcConfig) error {
+	c.Lock()
+	defer c.Unlock()
+
+	if len(c.configs) > 0 && config.AdvertiseName != c.configs[0].AdvertiseName {
+		return errors.New("Name mismatch")
```

Comment:
errors should start with a lowercase letter, also consider using `fmt.Errorf` (or the fancier error package) to include the values we're seeing. this could be frustrating to debug otherwise